### ISSUE TITLE
Hanging when evaluating many events 

### ISSUE THREAD

**Issue by ghost** [hunterhector/EvmEval#19]
Hello, I am trying to use @After evaluation with many events ~400 events in a document. I understand it does some sort of transitive closure but it took a very long time (not even finished). Do you suggest any idea to scale it? 

My project requirement is to evaluate a gold document with ~400 events (on @After) with the system output. All just one document.

**Comment by hunterhector** [comment]
I do remember the speed for temporal evaluation is rather slow. For your problem, it depends on what happened:

1. We adopt the TempEval code here for the core eval functions, the following: https://github.com/hunterhector/EvmEval/tree/master/evaluation_relations
Since we didn't write these, I do not know much about how to improve the speed.
1. If the hanging happen when we convert your data to the TempEval XML format, then it is likely a loop occurs. If you confirm that, then that should be a bug for us to fix.

**Comment by ghost** [comment]
Thanks for your prompt reply. 
It does convert it to the corresponding format. it definitely hangs on the core implementation of TempEval evaluation. 
Correct me if I am wrong. If we want to just evaluate the @after relation then it should be easy by just comparing all pairs from the reference (gold) (I mean all combinations of linked events)

So Precision and recall should be computed as follow: (excerpt from UzZaman paper) 
```
Precision = (# of system temporal relations that can be verified from reference 
annotation temporal closure graph / # of temporal relations in system output)

Recall = (# of reference annotation temporal relations that can be verified from system 
output’s temporal closure graph / # of temporal relations in reference
annotation) 
```
[UzZaman paper](https://www.aclweb.org/anthology/P11-2061.pdf)

**Comment by hunterhector** [comment]
Yes,  the **original** PR computation is like that. It is a mirror operation that checks the System graph and Gold graph for the opposite relations. But there is a problem in this method as described in his [thesis](https://urresearch.rochester.edu/institutionalPublicationPublicView.action?institutionalItemId=24810)(see section 6.1.6).

I haven't studied the UzZaman code deeply but I think it is possible to improve the speed. We didn't do that because we are afraid that our re-implementation could cause errors. I suppose that constructing the whole closure graph is slow, but a proper implementation should be polynomial, which shouldn't be implausible for 400 mentions. But that being said, rewriting the original code is quite tedious.

Another approach I could think of is to not create the closure graph. Instead, for every new relation we want to check, we can run a BFS on the other graph, to see if it is reachable. Though now for every new relation we can need to redo all computation, there are two potential benefits: a) we can at least see the progress b) since the number of relations could be much less than 400 (I assume), this may be acceptable.

However, be aware of the updated metric mentioned in the first paragraph. The formula is very complex in that case.

**Comment by ghost** [comment]
You are right, checking weather two-nodes can be reachable is more effective. Thanks a lot for pointing to the aforementioned thesis. The modified version in section 6.1.7  is actually what I want and it does address the problem I am thinking of. I really appreciate it.

My problem is easier where the system must output only one directed sequence of nodes.
